### PR TITLE
[Backend]: Configure Hikari DataSource for Database connection pooling

### LINKED ISSUE #554: [Backend] Use Hikari DataSource for Database connection pooling
In the backend application, currently, connection pooling is not enabled. [Hikari DataSource](https://github.com/brettwooldridge/HikariCP) should be used along with [Exposed](https://github.com/JetBrains/Exposed) in order to properly leverage DB connection pooling and performance of queries.

### PR DESCRIPTION
**Pull request by mrfamouskk7**
## Summary
In the backend application, currently, connection pooling is not enabled. [Hikari DataSource](https://github.com/brettwooldridge/HikariCP) should be used along with [Exposed](https://github.com/JetBrains/Exposed) in order to properly leverage DB connection pooling and performance of queries. Closes #554 

## A picture or screenshot regarding change
[Screencast from 10-10-22 06:06:54 PM IST.webm](https://user-images.githubusercontent.com/60615909/194867948-85d99c61-f558-4bbf-bb91-b0dfa59293f3.webm)

## Checklist

- [x] Build and linting is passing.
- [x] This change is not breaking existing flow of a system.
- [ ] I have written test case for this change.
- [x] This change is tested from all aspects.

### CODE REVIEW

**Comment by PatilShreyas** [comment]
@mrfamouskk7 can you please fix a conflict?

**Comment by mrfamouskk7** [comment]
> @mrfamouskk7 can you please fix a conflict?

Done

File: noty-api/application/build.gradle [hunk 1]
```diff
@@ -15,6 +15,7 @@ repositories {
 dependencies {
     // Data module
     implementation project(":data")
+    implementation "com.zaxxer:HikariCP:$hikariVersion"
```
Comment (PatilShreyas) [inline noty-api/application/build.gradle:18]: Why include this in the application module? I guess it's only needed in the data module.

File: noty-api/application/src/dev/shreyaspatil/noty/api/di/component/ConfigComponent.kt [hunk 2]
```diff
@@ -16,15 +16,15 @@
...
 import dev.shreyaspatil.noty.api.di.module.SecretKey
-import dev.shreyaspatil.noty.data.database.DatabaseConfig
 import javax.inject.Singleton
 
 @Singleton
 @Subcomponent(modules = [ConfigModule::class])
 interface ConfigComponent {
     @SecretKey fun secretKey(): String
-    fun databaseConfig(): DatabaseConfig
+    fun databaseConfig(): HikariConfig
```
Comment (PatilShreyas) [inline noty-api/application/src/dev/shreyaspatil/noty/api/di/component/ConfigComponent.kt:29]: Do not let the application module know about the Hikari config, so please revert this change. Let all DB-related config be in the `DatabaseConfig` data class only.

File: noty-api/data/src/dev/shreyaspatil/noty/data/database/NotyDatabase.kt [hunk 3]
```diff
@@ -25,17 +27,24 @@ import org.jetbrains.exposed.sql.transactions.transaction
...
-        password = databaseConfig.password
-    )
+    Database.connect(hikari(hikariConfig))
 
     transaction {
         SchemaUtils.createMissingTablesAndColumns(*tables)
     }
+}
+
+private fun hikari(hikariConfig: HikariConfig): HikariDataSource {
```
Comment (PatilShreyas) [inline noty-api/data/src/dev/shreyaspatil/noty/data/database/NotyDatabase.kt:40]: 1. Rename this function to `createDataSource`.
2. Change the return type to `DataSource` which is the super interface of all common Data Sources.

```suggestion
private fun createDataSource(databaseConfig: DatabaseConfig): DataSource {
```

**Review by PatilShreyas — CHANGES_REQUESTED** [review]
@mrfamouskk7 Thanks for raising a PR and attaching a screenshot as well. It really makes easy to review the changes.

File: noty-api/application/build.gradle (hunk 1 above)
Comment (mrfamouskk7) [reply noty-api/application/build.gradle]: Was using HikariConfig in Application module but it was not necessary so removed it.

Comment (mrfamouskk7) [reply noty-api/application/build.gradle]: Was using HikariConfig in Application module but it was not necessary so removed it.

File: noty-api/application/src/dev/shreyaspatil/noty/api/di/component/ConfigComponent.kt (hunk 2 above)
Comment (mrfamouskk7) [reply noty-api/application/src/dev/shreyaspatil/noty/api/di/component/ConfigComponent.kt]: Changed it to `DatabaseConfig` :+1:

File: noty-api/data/src/dev/shreyaspatil/noty/data/database/NotyDatabase.kt (hunk 3 above)
Comment (mrfamouskk7) [reply noty-api/data/src/dev/shreyaspatil/noty/data/database/NotyDatabase.kt]: Renamed it as per suggestion.

File: noty-api/application/src/dev/shreyaspatil/noty/api/di/component/ConfigComponent.kt (hunk 2 above)
Comment (mrfamouskk7) [reply noty-api/application/src/dev/shreyaspatil/noty/api/di/component/ConfigComponent.kt]: Changed it to `DatabaseConfig` :+1:

File: noty-api/data/src/dev/shreyaspatil/noty/data/database/NotyDatabase.kt [hunk 4]
```diff
@@ -16,26 +16,37 @@
...
+    val config = HikariConfig()
+    with(databaseConfig) {
+        config.driverClassName = driver
+        config.password = password
+        config.jdbcUrl = "jdbc:postgresql://$host:$port/$name"
+        config.maximumPoolSize = maxPoolSize
+        config.username = user
+    }
+    config.validate()
+    return HikariDataSource(config)
```
Comment (PatilShreyas) [inline noty-api/data/src/dev/shreyaspatil/noty/data/database/NotyDatabase.kt:51]: This now looks good

**Review by PatilShreyas — APPROVED** [review]
Thanks, @mrfamouskk7, change looks good to me 😄. Merging it 🚀

**Comment by mrfamouskk7** [comment]
> Thanks, @mrfamouskk7, change looks good to me smile. Merging it rocket

Thanks @PatilShreyas for the opportunity :bow: ...  It was really interesting to work on the setup :smiley:
